File: Apple Capture/src/server/Services/AppleGenerationService.lua
local Players = game:GetService('Players');

local Knit = require(game:GetService('ReplicatedStorage').Knit);
local Maid = require(Knit.Util.Maid);
local RoundService;

local newVector3 = Vector3.new;
local mathRandom = math.random;

local applePrefab = game:GetService('ServerStorage').Apple;
local appleHalfHeight = applePrefab.Size.Y * 0.5;

local baseplate = workspace.Baseplate;
local baseplateHalfWidth = baseplate.Size.X * 0.5;
local baseplateHalfHeight = baseplate.Size.Y * 0.5;

local maid = Maid.new();

local AppleGenerationService = Knit.CreateService({
    Name = 'AppleGenerationService',
    AppleCollectors = {},
    Client = {}
});

function AppleGenerationService:ResetAppleCollectors()
    self.AppleCollectors = {};
end;

function AppleGenerationService:StartGeneration()
    for i = 1, 45 do
        local collected = false;
        local apple = applePrefab:Clone();
        local applePos = newVector3(mathRandom(-baseplateHalfWidth, baseplateHalfWidth), baseplateHalfHeight + appleHalfHeight, mathRandom(-baseplateHalfWidth, baseplateHalfWidth));
        apple.Position = applePos;
        apple.Parent = workspace;
        apple.Touched:Connect(function(instance)
            if ((instance.Position - apple.Position).Magnitude > 5) or (collected) then return; end;
            local plr = Players:GetPlayerFromCharacter(instance.Parent);
            if not plr then return; end;
            collected = true;
            apple:Destroy();
            local collectedApples = self.AppleCollectors[plr];
            collectedApples = collectedApples and collectedApples + 1 or 1
            self.AppleCollectors[plr] = collectedApples;
            RoundService:AppleCollected(plr, collectedApples);
        end);
        maid:GiveTask(apple);
    end;
end;

function AppleGenerationService:StopGenerationAndClearApples()
    maid:DoCleaning();
end;

function AppleGenerationService:KnitStart()

end;

function AppleGenerationService:KnitInit()
    RoundService = Knit.Services.RoundService;
end;

return AppleGenerationService;

File: AppleCaptureTS/src/server/services/apple-generation-service.ts
import { KnitServer as Knit, Maid } from '@rbxts/knit';
import { Players, ServerStorage, Workspace } from '@rbxts/services';

let RoundService: typeof Knit.Services.RoundService;

const applePrefab = ServerStorage.FindFirstChild('Apple') as BasePart;
const appleHalfHeight = applePrefab.Size.Y * 0.5;

const baseplate: BasePart = Workspace.FindFirstChild('Baseplate') as BasePart;
const baseplateHalfWidth: number = baseplate.Size.X / 2;
const baseplateHalfHeight: number = baseplate.Size.Y / 2;

declare global {
    interface KnitServices {
        AppleGenerationService: typeof AppleGenerationService;
    }
}

const AppleGenerationService = Knit.CreateService({
    Name: 'AppleGenerationService',

    appleCollectors: new Map<Player, number>(),
    maid: new Maid(),

    Client: {},

    resetAppleCollectors() {
        this.appleCollectors.clear();
    },

    startGeneration() {
        for (let i = 0; i < 45; i++) {
            let collected: boolean = false;
            let apple: BasePart = applePrefab.Clone();
            let applePos: Vector3 = new Vector3(math.random(-baseplateHalfWidth, baseplateHalfWidth), baseplateHalfHeight + appleHalfHeight, math.random(-baseplateHalfWidth, baseplateHalfWidth));
            apple.Position = applePos;
            apple.Parent = Workspace;
            apple.Touched.Connect((instance: BasePart) => {
                if ((instance.Position.sub(apple.Position).Magnitude > 5) || (collected)) return;
                let plr: Player | undefined = Players.GetPlayerFromCharacter(instance.Parent);
                if (!plr) return;
                collected = true;
                apple.Destroy();
                let collectedApples: number | undefined = this.appleCollectors.get(plr);
                collectedApples = collectedApples ? collectedApples + 1 : 1;
                this.appleCollectors.set(plr, collectedApples);
                RoundService.appleCollected(plr, collectedApples);
            });
            this.maid.GiveTask(apple);
        }
    },

    stopGenerationAndClearApples() {
        this.maid.DoCleaning();
    },

    KnitInit() {
        RoundService = Knit.Services.RoundService;
    }
});

export = AppleGenerationService;
